### ISSUE TITLE
`set_buffer_limit` API is confusing, especially regarding the value 0.

### ISSUE THREAD

**Issue by briansmith** [ctz/rustls#716]
If we're going to support an unbounded buffered mode, then I think that should be requested by `set_buffer_limit(None)` or `set_buffer_limit(BufferSize::Unbounded)` or similar, not by `set_buffer_limit(0)`. `set_buffer_limit(0)` should mean "no buffering." I imagine that some applications are likely to want to have a zero-size buffer. In particular, TCP already has buffering within the socket at the OS layer, so some TCP-based applications will not want to have two buffers. Thus we should support 0 as a real limit instead of its current meaning of "unlimited."

The documentation also needs to be improved.

**Comment by briansmith** [comment]
Currently (as of yesterday in the main branch) the buffer limit is 64K. That's a good improvement over the previous default of "unbounded". However, why 64K?

AFAICT, the main benefit of Rustls doing this benefit occurs when an applications send a large message that would be split into multiple records, where Rustls will encrypt them all and then send them in bulk in a single syscall; i.e. the primary benefit seems to be reducing syscall overhead for large messages. Are there other significant benefits to this buffering?

I think a lot of applications primarily send small messages and would benefit more from better backpressure, so 0 makes more sense to me as the default. I do think that especially for TCP-based applications, controlling buffering at the socket layer makes more sense.

**Comment by djc** [comment]
> I think a lot of applications primarily send small messages and would benefit more from better backpressure, so 0 makes more sense to me as the default.

Are you saying we should change the `process_new_packets()` API to take a `Write` parameter which we can write into immediately? With the current API, where I/O is separate from record processing, it doesn't seem possible to have a buffer limit at 0 (assuming you mean actual 0 rather than sentinel-`None`).

**Comment by briansmith** [comment]
> With the current API, where I/O is separate from record processing, it doesn't seem possible to have a buffer limit at 0 (assuming you mean actual 0 rather than sentinel-`None`).

What I expected to happen when I set the buffer limit to `0` is to get the smallest buffer I can get, and as far as possible minimize double-buffering.

**Comment by ctz** [comment]
> The documentation also needs to be improved.

Well I think this is something we agree on.

The function and purpose of this buffering was initially to prevent having to commit to building an IO abstraction.  Now this buffering is primarily prevent the ever-growing complexity of the "rust IO ecosystem" and the lack of generally-accepted IO abstractions and async scheduler leaking into this library. The cost of that is outgoing data is buffered within the library; given that, a zero-sized buffer is not a coherent thing to ask for.

From this it sounds like you want rustls to be "write-through", which AFAICT will involve finding or building an abstraction that covers blocking sockets, non-blocking sockets, IOCP, and each mutually-incompatible async IO ecosystem.

**Comment by djc** [comment]
> From this it sounds like you want rustls to be "write-through", which AFAICT will involve finding or building an abstraction that covers blocking sockets, non-blocking sockets, IOCP, and each mutually-incompatible async IO ecosystem.

We could instead use buffers allocated by the caller. I think doing `fn process_new_packets(input: &mut [u8], output: &mut [u8])` (without the use of connection-internal buffers) would be relatively straightforward, and would allow different abstractions to optimize their own buffering.

**Comment by ctz** [comment]
There would also have to be an output buffer for `std::io::Write::write` since that is the principal source of TLS frames.

**Comment by djc** [comment]
Ah, true. Could define a struct for it, somewhat like this:

```rust
struct TlsIo<'a> {
     incoming_tls: &'a mut [u8],
     incoming_app_data: &'a mut [u8],
     outgoing_tls: &'a mut [u8],
}
```

**Comment by briansmith** [comment]
I propose the following course of action:

1. Stop using `0` as a sentinel that means "use the default"; instead just expose a `BUFFER_DEFAULT_SIZE` constant that can be used as the argument if the default is desired.
2. Improve the documentation.
3. Release
4. Work together to design a new I/O interface between Rustls and the application and the OS that minimizes copies and buffers and is optimized for performance and low memory overhead. This would be a big project and the design would drive a lot of the optimization work from here forward.

WDYT?

**Comment by djc** [comment]
Works for me, though I'm also fine if we want to do the design work first. A lot of the implementation is already done.

Would be nice to have a release that contains #544 so that we can release a Quinn version implementing RFC 9000.

**Comment by djc** [comment]
I believe the core issue here has been resolved in 4ff8bc156d991312f4dd9b75fac93f6c254b68cf. I suppose we could close this issue and discuss a new I/O design elsewhere?

**Comment by djc** [comment]
Let's close this. Other issues can be discussed in #668, for example.
